### PR TITLE
Add tokenprices tables

### PR DESCRIPTION
**Pull request by antonyip**
This will ease the load on the snowflake database

### CODE REVIEW

File: models/tokenprice/tokenprice_dfkgold.sql [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    )
+}}
+
+with
+
+historical_prices as (
+    select
+        date_trunc('day',timestamp) as timestamp,
+        avg(price) as price
+    from harmony.dev.ant_token_prices
```
Comment (chuxinh) [inline models/tokenprice/tokenprice_dfkgold.sql:17]: better not ref `dev` here

Comment (chuxinh) [inline models/tokenprice/tokenprice_dfkgold.sql]: better not ref `dev` here

File: models/tokenprice/tokenprice_dfkgold.sql [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    where {{ incremental_load_filter_withdays("timestamp", -5) }}
+    and token_address = '0x3a4edcf3312f44ef027acfd8c21382a5259936e7' -- gold
```
Comment (chuxinh) [inline models/tokenprice/tokenprice_dfkgold.sql]: indent

File: models/tokenprice/tokenprice_dfkgold.sql [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        tags=['tokenprice'],
+        cluster_by=['timestamp']
+    )
+}}
+
+with
+
+historical_prices as (
+    select
+        date_trunc('day',timestamp) as timestamp,
```
Comment (chuxinh) [inline models/tokenprice/tokenprice_dfkgold.sql]: call it `day`? or `event_date`

File: models/tokenprice/tokenprice_dfkgold.sql [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+final as (
+    select 
+        timestamp,
+        avg(price) as price
+    from combine
+    group by 1
+)
+
+select * from final
```
Comment (chuxinh) [inline models/tokenprice/tokenprice_dfkgold.sql]: add extra line

File: models/tokenprice/tokenprice_dfkgold.sql [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    select 
+        block_timestamp,
+        java_hextoint(substr(data,3+64*0,64)) as amount0In,
+        java_hextoint(substr(data,3+64*1,64)) as amount1In,
+        java_hextoint(substr(data,3+64*2,64)) as amount0Out,
+        java_hextoint(substr(data,3+64*3,64)) as amount1Out
+    from {{ ref('logs') }}
+    where {{ incremental_load_filter_withdays("block_timestamp", -5) }}
+    and topics[0] = '0xd78ad95fa46c994b6551d0da85fc275fe613ce37657fb8d5e3d130840159d822' -- Swap
+    and evm_contract_address = '0x321eafb0aed358966a90513290de99763946a54b' -- gold/jewel Pool
```
Comment (chuxinh) [inline models/tokenprice/tokenprice_dfkgold.sql]: indent

File: models/tokenprice/tokenprice_dfkgold.sql [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+),
+
+daily_trades as (
+    select 
+        date_trunc('day', block_timestamp) as ddate,
```
Comment (chuxinh) [inline models/tokenprice/tokenprice_dfkgold.sql]: typo

File: models/tokenprice/tokenprice_dfkgold.sql [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        sum(amount0In) as s0i,
+        sum(amount1In) as s1i,
+        sum(amount0Out) as s0o,
+        sum(amount1Out) as s1o
```
Comment (chuxinh) [inline models/tokenprice/tokenprice_dfkgold.sql]: can this be more informative?

File: models/tokenprice/tokenprice_dfkgold.sql [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    from raw_logs
+    group by 1
+),
+
+daily_trades_prices as (
+    select
+        ddate as timestamp,
+        1/ ((s0i + s0o) / (s1i+s1o)) * pow(10,-15) * j.price as price
```
Comment (chuxinh) [inline models/tokenprice/tokenprice_dfkgold.sql]: i'm so confused about the calculation here 😵

File: models/tokenprice/tokenprice_dfkgold.sql [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    from daily_trades d
+    left join harmony.dev.ant_view_token_price_jewel j on j.timestamp = d.ddate
```
Comment (chuxinh) [inline models/tokenprice/tokenprice_dfkgold.sql]: `on` to a new line. 
also not recommended to use `dev`

File: models/tokenprice/tokenprice_dfkgold.sql [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+historical_prices as (
+    select
+        date_trunc('day',timestamp) as timestamp,
+        avg(price) as price
+    from harmony.dev.ant_token_prices
+    where {{ incremental_load_filter_withdays("timestamp", -5) }}
+    and token_address = '0x3a4edcf3312f44ef027acfd8c21382a5259936e7' -- gold
+    group by 1
+    order by 1 desc
```
Comment (chuxinh) [inline models/tokenprice/tokenprice_dfkgold.sql]: maybe no need to order here as you do ordering in the final query using the cluster by logic above

**Review by chuxinh — CHANGES_REQUESTED** [review]
just left comment for first query as the others are having similar structure
maybe it's also better to add comments in terms of how the price is calculated and also avoid using dev in productionised model

File: macros/func_hextoint.sql [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+as $$
+import java.math.*;
+class MyClass {
+  public static String x(String s)
+  {
+    return new BigInteger(s,16).toString();
+  }
+}
+$$;
+{%- endmacro %}
```
Comment (chuxinh) [inline macros/func_hextoint.sql:16]: add line

Comment (chuxinh) [inline macros/func_hextoint.sql]: add line

File: macros/func_hextoint.sql [hunk 12]
```diff
@@ -0,0 +1,16 @@
+
+{% macro create_java_to_hex() -%}
+create or replace function java_hextoint (s string)
```
Comment (chuxinh) [inline macros/func_hextoint.sql]: also add `{{ target.database }}.{{ target.schema }}`?
ref: https://discourse.getdbt.com/t/using-dbt-to-manage-user-defined-functions/18

File: macros/func_hextoint.sql [hunk 13]
```diff
@@ -0,0 +1,16 @@
+
```
Comment (chuxinh) [inline macros/func_hextoint.sql]: also would recommend creating `macros/udfs/`?

File: macros/incremental_utils.sql [hunk 14]
```diff
@@ -6,3 +6,11 @@
       true
     {% endif %}
 {%- endmacro %}
+
+{% macro incremental_last_x_days(time_col, time_in_days) -%}
+    {% if is_incremental() %}
+        {{ time_col }} >= current_date() - {{ time_in_days }}
```
Comment (chuxinh) [inline macros/incremental_utils.sql]: i'd recommend still using the max timestamp from the table so 
```suggestion
        {{ time_col }} >= select max({{ time_col }}) from {{ this }} - interval '{{ time_in_days }} day'
```

File: macros/incremental_utils.sql [hunk 15]
```diff
@@ -6,3 +6,11 @@
...
+    {% else %}
+        true
+    {% endif %}  
+{%- endmacro %}
```
Comment (chuxinh) [inline macros/incremental_utils.sql]: add line

File: models/tokenprice/tokenprice_dfkgold.sql [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+    )
+}}
+
+with
+
+jewel_price as (
+    select
+        *
+    from {{ ref("tokenprice_jewel") }}
+    where {{ incremental_last_x_days("timestamp", 3) }}
```
Comment (chuxinh) [inline models/tokenprice/tokenprice_dfkgold.sql:17]: any specific reason for choosing 3 days here?

File: models/tokenprice/tokenprice_jewel.sql [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+    select 
+        *
+    from daily_trades_prices
+),
+
+final as (
+    select 
+        timestamp,
+        avg(price) as price
+    from combine
```
Comment (chuxinh) [inline models/tokenprice/tokenprice_jewel.sql:61]: think you can directly use `daily_trades_prices`

**Review by chuxinh — APPROVED** [review]
LGTM after nits
Curious why you're breaking all token prices into separate tables instead of combining into one since the logics are quite similar but from different logs as i can tell

File: models/tokenprice/tokenprice_dfkgold.sql (hunk 16 above)
Comment (antonyip) [reply models/tokenprice/tokenprice_dfkgold.sql:17]: dbt runs in the middle of the day, so 2 days should be safe (so we cover all the edge cases when updating tables), but I'm not 100% sure about timezones. so left it to 2+1.

**Comment by antonyip** [comment]
> LGTM after nits Curious why you're breaking all token prices into separate tables instead of combining into one since the logics are quite similar but from different logs as i can tell

Plan is to have per minute price data ingested in the future.. Having it all in 1 table will make it heavy on the database. Also, the general structure is the same, but the details are quite hard to generalize (different decimals for tokens , different formulas for different token pairs, sometimes its token0/token1, sometimes its token1/token0 ) and I feel this will cause quite a heavy compute in my 20 dfk_tables.

File: macros/func_hextoint.sql (hunk 12 above)
Comment (antonyip) [reply macros/func_hextoint.sql]: gave me an error when I tried to do 
`create or replace function {{ target.database }}.{{ target.schema }}.java_hextoint (s string)`
not too sure if its a syntax problem or a macro problem.

File: macros/incremental_utils.sql (hunk 14 above)
Comment (antonyip) [reply macros/incremental_utils.sql]: also gave me an error
```
2022-01-30 00:28:54.264179 (Thread-3): Database Error in model metric_miner_count (models/metrics/metric_miner_count.sql)
  001003 (42000): SQL compilation error:
  syntax error line 14 at position 27 unexpected 'select'.
  syntax error line 14 at position 38 unexpected 'block_timestamp'.
  syntax error line 17 at position 1 unexpected ','.
```
compiled dbt statement looks like this
```
      create or replace temporary table HARMONY.DEV.metric_gas_used_count__dbt_tmp  as
      (select * from(
            

with

daily as (

    select 
        date_trunc('day', block_timestamp) as metric_date,
        'daily' as metric_period,
        sum(gas_used) as gas_count
    from HARMONY.DEV.blocks
    where 
        block_timestamp >= select max(block_timestamp) from HARMONY.DEV.metric_gas_used_count - interval '3 day'
    
    group by 1

),

hourly as (

    select 
        date_trunc('hour', block_timestamp) as metric_date,
        'hourly' as metric_period,
        sum(gas_used) as gas_count
    from HARMONY.DEV.blocks
    where 
        block_timestamp >= select max(block_timestamp) from HARMONY.DEV.metric_gas_used_count - interval '3 day'
    
    group by 1

),

minute as (

    select 
        date_trunc('minute', block_timestamp) as metric_date,
        'minute' as metric_period,
        sum(gas_used) as gas_count
    from HARMONY.DEV.blocks
    where 
        block_timestamp >= select max(block_timestamp) from HARMONY.DEV.metric_gas_used_count - interval '3 day'
    
    group by 1

),

final as (
    select * from daily
    union all
    select * from hourly
    union all
    select * from minute
)

select * from final
            ) order by (metric_date)
      );
      
```

File: models/tokenprice/tokenprice_jewel.sql (hunk 17 above)
Comment (antonyip) [reply models/tokenprice/tokenprice_jewel.sql:61]: Plan is to have hourly data in the future. Leaving the structure intact for future changes.

**Pull request merged**